### PR TITLE
Re-enable imageUpload native integration

### PR DESCRIPTION
**Pull request by fdocr**
## What type of PR is this? (check all applicable)

- [x] Bug Fix
- [x] Optimization

## Description

We had disabled this native integration as we rolled out a relatively big refactor. We were blocked by the Apple review, and given the many moving pieces of the refactor we decided to go ahead with rolling out the web (`forem/forem`) parts of it and re-enable this integration whenever the iOS app update becomes available on the AppStore.

[More details here](https://github.com/forem/forem/pull/15212#discussion_r755507405)

## Related Tickets & Documents

N/A

## QA Instructions, Screenshots, Recordings

1. Using the (latest version) iOS mobile app add `canary1.forem.wtf` instance
2. Sign in and open the Post editor
3. Upload a cover image
4. Upload an image to use in the markdown
5. Both should work as expected using the native image picker

Example: https://canary1.forem.wtf/fdoxyz/mobile-image-picker-fpb

### UI accessibility concerns?

None: UI remains the same but we're using the native image picker instead of the fully web solution

## Added/updated tests?

- [x] Yes

## [Forem core team only] How will this change be communicated?

- [x] This change does not need to be communicated, and this is why not: The app remains unchanged, except for re-enabling a feature that enhances the image picker on mobile apps.

## [optional] What gif best describes this PR or how it makes you feel?

WWE undertaker - back to life

### CODE REVIEW

**Comment by citizen428** [comment]
Hi @fdoxyz, what's the status of this PR, are we still waiting for the Apple review?

**Comment by fdocr** [comment]
Cover image upload seems to work well. Last week I found that the image upload (from the markdown toolbar) fails because we don't have a selection or the cursor isn't positioned (or something along those lines).

Screen Shot 2022-02-28 at 10 55 07

I'm thinking that since we're on mobile, the moment the user taps on the "upload image" icon the editor's text area loses focus (maybe?). I didn't make much progress in fixing this last week, but maybe someone from @forem/frontend may have a suggestion for how to solve this?

Failures are happening in [textAreaUtils.js](https://github.com/forem/forem/blob/main/app/javascript/utilities/textAreaUtils.js#L233-L246) and [handleImageUploadEnd](https://github.com/forem/forem/blob/main/app/javascript/crayons/MarkdownToolbar/MarkdownToolbar.jsx#L243-L279). Maybe giving the `selectionStart` and likely `selectionEnd` default values would be a good solution? i.e. always append if no selection exists

**Comment by fdocr** [comment]
Thanks a lot @aitchiss!

> I have no idea right now why this would be at all different in the mobile app... also this function is currently being used in the live version of the app without issue, right? In terms of inserting the new markdown nothing has actually changed 🙃

Yes, I agree this is the reason why it's odd, since nothing seems to have changed at first glance. Right now the default JS flow is what users face on mobile, which works as expected. UI looks like this:

Screen Shot 2022-03-01 at 09 55 03

This is what it should look like when using this PR's branch (native image picker is presented):

Screen Shot 2022-03-01 at 09 57 48

> If you can help me get set up locally I'd be happy to look into this a bit further - I'm sure we can get to the bottom of it!

You're on macOS, right? I remember you once mentioned you had the simulator running locally. These are the steps to test this:

1. Boot `forem/forem-ios` from the `main` branch on XCode and run it using the simulator
   - Choose the simulator device (i.e. `iPhone 13 mini`) from the drop down and then `CMD + R`
3. Add `fdoxyz.ngrok.io` or a Canary Instance to the app on the simulator
   - I'm asking `#systems` to deploy this branch on a canary instance so you can try it whenever you need
4. Open Safari and go to `Develop` > `Simulator` > `Forem instance` to open the console
   - If `Develop` isn't visible you'll need to go to Safari preferences to enable dev tools 
   - The reason we need the simulator is because production apps won't allow this console access on them
5. Try the editor image upload (both cover & markdown toolbar) and errors should appear in that dev tools console

Some more thoughts/context on this:
- I just recently started running in to some CSRF errors when uploading the image
   - I didn't saw this happen until today while going over the debug instructions from above 🤔 
   - I'm trying out a few options to sort this out like [this PR](https://github.com/forem/ForemWebView-ios/pull/49)
   - I hope to get all of this sorted so you can debug this using `main` (on `forem/forem-ios`) starting tomorrow along with the Canary instance
- I'm pushing some very small changes that add some `console.log` to the FE code for easier debugging on the simulator/Safari dev tools
- Let me know if anything from the instructions above isn't clear 🙂

**Comment by aitchiss** [comment]
Yay thanks for all the steps to get this running! I'll look into this tomorrow ❤️

File: app/javascript/article-form/components/ArticleCoverImage.jsx
Comment (fdocr) [inline app/javascript/article-form/components/ArticleCoverImage.jsx:114]: Temporary log for debugging on Safari Dev Tools

File: app/javascript/article-form/components/ImageUploader.jsx
Comment (fdocr) [inline app/javascript/article-form/components/ImageUploader.jsx:337]: Temporary log for debugging on Safari Dev Tools

**Comment by aitchiss** [comment]
So like we were saying before, it seems the root issue is that `textArea` is not defined when it's evaluated in `handleImageUploadStarted`. I tried evaluating some other variables that functions should have access to in the debugger and _funnily enough_ all of them were undefined. I think this is a JS `this` issue 😵  @fdocr - can you try changing the `handleImageUploadStarted` (and I guess `handleImageUploadEnd`) to be e.g.

```js
function handleImageUploadStarted() {
...
}
```

instead of an arrow function?

**Comment by fdocr** [comment]
So after some more debugging I wasn't able to get it working despite the suggestions @aitchiss. I just pushed the last one and these are some of the things I'm noticing:

First, the textArea is still `null` when I use [a breakpoint here](https://github.com/forem/forem/pull/15487/files#diff-6e476aa868e6891c12d1ce12d9d685865c544ebb4628a975224a4c3d67d9aee9R223) with both the function and arrow style functions (screenshot below)

Screen Shot 2022-03-04 at 09 15 15

I have a new hypothesis why this is happening (thinking out loud bc I'm not very certain about this):

I left a breakpoint right at the time the `render` is being executed, where the textArea is also `null`. When the editor _first loads_ we render a "minimal UI" where the textArea is `null`. The `render` function gets rendered and the textArea is `null` at that point:

Screen Shot 2022-03-04 at 09 25 01

Then, when the actual editor loads properly we do have the textArea available on that variable:

Screen Shot 2022-03-04 at 09 28 18

So maybe the Toolbar or the corresponding function was left with a `null` reference from the first time it gets rendered?

One thing to note too is that whenever the native image picker gets displayed on screen (overlays the webview with a modal view) the `render` function gets called again. This is a call that gets executed by the webview/JS context and not triggered by the app itself AFAICT. I'm very unsure if this is a problem, but I don't think so since that `render` should be idempotent (no issues should arise if it gets rendered again), right?

> and for some reason if I run the app on an ngrok url I can't add it in the Forem app 🤷‍♀️ 

That's odd. My local instance works well with ngrok. Any specific error you see displayed in the app's UI when trying to add it?

I'm happy to share my local instance for a quicker iteration/debugging if you'd like. Just let me know either today or next week whenever you have some time 🙂

**Comment by aitchiss** [comment]
> That's odd. My local instance works well with ngrok. Any specific error you see displayed in the app's UI when trying to add it?

This is the error I get:
we couldn't find this community domain


It seems like when the app calls `/api/instance` to get the info, it gets `nil` back and errors with `"Response could not be decoded because of error:\nThe data couldn’t be read because it is missing."`... when I visit the same URL in the browser I get the JSON metadata:

```json
{
  "context": null,
  "cover_image_url": "https://1eee-2a00-23c8-8686-4c01-b13e-e7e6-8c1c-e461.ngrok.io/assets/social-media-cover.png",
  "description": null,
  "display_in_directory": true,
  "domain": "1eee-2a00-23c8-8686-4c01-b13e-e7e6-8c1c-e461.ngrok.io",
  "logo_image_url": "https://1eee-2a00-23c8-8686-4c01-b13e-e7e6-8c1c-e461.ngrok.io/assets/icon.png",
  "name": "MyLocal",
  "tagline": null,
  "version": "edge.20220304.0",
  "visibility": "public"
}
```

I'm sure it's something really basic I'm missing but I don't see it 🙈

**Comment by aitchiss** [comment]
TBH I am starting to feel like the debugger is giving me the run around 😂  Everything I try to log out the value of reports the same reference error in the debugger, e.g. even the `getSecondaryFormattingButtons` function, even though I can see them plainly in the UI 🤷‍♀️  Definitely would love to get this working with my local instance to poke some more!

**Comment by fdocr** [comment]
Interesting to see ngrok now using UUID generated subdomains! Those are much longer than the hex values they used to use before 😓 

I think that even free accounts can use custom subdomains, what they can't do is "reserve them", but I may be wrong... Just a tiny and entirely optional suggestion in case that may help make things easier 🙂 

```bash
# https://ngrok.com/docs#http-subdomain
ngrok http -subdomain=my-forem 3000
```


> I'm sure it's something really basic I'm missing but I don't see it 🙈

Looking at that error I think this may be caused by the description/tagline being `null`.

We're moving away from that API call soon though (https://github.com/forem/forem-ios/issues/130), so for now I think maybe adding those two values on the local instance (`/admin` dashboard configs) with anything so they're not `null` might help, until we get that change merged 😅 

> Definitely would love to get this working with my local instance to poke some more!

Cool, if you need any help from my end let me know! 😊

**Comment by aitchiss** [comment]
> Looking at that error I think this may be caused by the description/tagline being null.

Unfortunately setting them didn't help, but I did just comment out the validation around Forem metadata and managed to fudge it so my ngrok url could be added 😄 

I'll take it for a spin on Monday and see if I have any more joy debugging this! Thanks for the subdomain tip!!

**Comment by aitchiss** [comment]
Hey @fdocr I've pushed up a change! The more I thought about it the more I thought it was just _strange_ that I was storing a DOM element in state... sure enough, that seems to be the root of our issues. I've updated it to `useRef` instead, and we're no longer seeing these strange `textArea` being null problems.

The image upload still isn't working though 🙃  It now gets as far as showing the placeholder `Uploading image`, then I see an error snackbar (below),  and the `handleImageUploadEnd` doesn't seem to be called by either the error or success handler.

response could not be serialized

**Comment by fdocr** [comment]
This is awesome thank you @aitchiss!

I just tried it and it seems to work perfectly for me locally and on Canary 1. The changes you pushed here are now live on that canary instance (`document.body.dataset.latestCommitId` returns `43a26c0` on there).

[This is the post](https://canary1.forem.wtf/fdoxyz/mobile-image-picker-fpb) I was able to upload with the native image picker 🙌 

> The image upload still isn't working though 🙃 It now gets as far as showing the placeholder Uploading image, then I see an error snackbar (below), and the handleImageUploadEnd doesn't seem to be called by either the error or success handler.

I think the problem you experienced may have been caused by something in the local environment setup. Maybe you already did, but asking just in case: Did you have both `APP_DOMAIN="XYZ.ngrok.io"` and `APP_PROTOCOL="https"` on your `.env` file? I've seen image uploads fail unless both of these are specified, though it may be something else causing it.

Anyways, in the mean time I'm going to merge `main` and have a look at the failing specs to finally open for review.

**Comment by aitchiss** [comment]
> I think the problem you experienced may have been caused by something in the local environment setup. Maybe you already did, but asking just in case: Did you have both APP_DOMAIN="XYZ.ngrok.io" and APP_PROTOCOL="https" on your .env file? I've seen image uploads fail unless both of these are specified, though it may be something else causing it.
>
>Edit/Note: A new app version (update) was released to the store last Friday. So it may be worth trying to pull the most recent main if you're running the XCode-Simulator to test this 🙂

Hmm it's definitely possible I didn't have the `APP_PROTOCOL` set properly when I was testing with ngrok... I just tried on the simulator (latest code from `main`), pointing at Canary 1 and I still see the same though 🤷‍♀️  It seems like this is very likely a local issue with the simulator, so I'm happy to defer to folks able to test on real devices 🙂

**Review by aitchiss — APPROVED** [review]
I mentioned on the main thread I had some issues testing this on the simulator, but the integration with the toolbar is definitely working, and there's no regression in the main app ✨ 

Approving on the basis of others being able to do the manual testing on iOS devices 🙂

**Review by joshpuetz — APPROVED** [review]
Testing and working in latest Forem-ios on device!

**Pull request merged**
